rfcs/0019-storing-client-data.md

```diff
@@ -0,0 +1,660 @@
+
+let write_transaction sql txn account_before_sending_txn =
+  write_transaction_with_no_commit sql txn account_before_sending_txn None
+  Sql.commit sql
+```
+
+The archive process will similarly persist a block along with its transactions whenever it receives the `Breadcrumb_added` diff. Additionally, we will compute the `receipt_chain_hash` of each `user_command` in the block using `Breadcrumb_added.sender_receipt_chains_from_parent_ledger`. This new computed `receipt_chain_hash` along with its parent `receipt_chain_hash` will be added to the `Receipt_chain_datbase` Below is psuedocode for persisting a block:
+
+```ocaml
+(* filtered transition only contains transitions that we are interested in following *)
+let write_block (~sender_receipt_chains_from_parent_ledger: Receipt_chain_hash.t Public_key.Compressed.Map.t) ~sql ~receipt_chain_database filtered_transition_with_hash previous_ledger date =
+  let {With_hash.data=filtered_transition; hash=state_hash} = filtered_transition_with_hash in
+  let transactions = Staged_ledger_diff.transactions @@  External_transition.staged_ledger_diff filtered_transition in
+  let saved_block = {
+    state_hash = State_hash.to_string state_hash;
+    creator = Staged_ledger_diff.proposer (External_transition.staged_ledger_diff)
+    status=Pending
+    ...
+  } in
+  Sql.write sql['block'] saved_block;  
+  List.iter transactions ~f:(fun transaction ->
+    (* Add receipt chain hashes*)
+    (match transaction with
+    | User_command user_command ->
+      let previous = Option.value_exn (Ledger.get sender_receipt_chains_from_parent_ledger transaction.sender) in
+      Receipt_chain_database.add ~previous receipt_chain_database transaction
+    | _ -> ());
+    let previous_sender_account = Option.value_exn (Ledger.get_account ledger location) in
+    write_transaction_with_no_commit ~time_first_seen:date sql ~is_manually_added:false transaction previous_sender_account transition_with_hash
+  )
+  Sql.commit sql
+```
+
+When the archive process receives the `Root_transitioned` diff, it will mainly update the `consensus_status` of a block. Below is the psuedocode for it:
+
+```ocaml
+let update_status sql finalized_transition deleted_transitions =
+  Sql.update sql['block'] finalized_transition.hash Confirmed;
+  List.iter deleted_transitions ~f:(fun deleted_transition ->
+    Sql.update sql['block'] deleted_transition.hash Failed;
+  );
+  Sql.commit sql
+```
+
+#### Client Queries
+
+When a client wants to make a query, it will talk to an API process. The API process will invoke a server that makes calls to the daemon process and the database. The API process will talk to the daemon through Graphql. The API process will make direct SQL calls to the database and present it to the client. The following subsections will describe queries how to use the databases to make complicated queries:
```

Comment:
We said the API would be part of the daemon process, right? Can you update this section to reflect that? (you can say it's an API module inside of the daemon process)

Specifically here's what I thought we were doing:
API module is part of the daemon process. It API process can directly hook into data-structures in the daemon as required and will make direct SQL reads to the database. (the API treats the sql database as read-only for all the use-cases I see in this doc)